Staticize CASCADE_PROPERTIES, (temporarily) fix stylo path for animations, and introduce the long-term path to follow

emilio opened the pull request:
<!-- Please describe your changes on the following line: -->

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

<!-- Either: -->
- [x] These changes do not require tests because refactoring + geckolib

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

@bholley: I was going to do this "the good way", but that involves quite a few properties, so I thought I'd rather unlock stylo before :)

r? @bholley

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/11972)

<!-- Reviewable:end -->

emilio committed:
style: Make CASCADE_PROPERTY a real static array.

Why wasn't this done before?

emilio committed:
style: Simplify generation of need_clone methods in servo

emilio committed:
style: Add a temporary workaround for geckolib animations.

With needs_borrow we will be able to remove this, though we'll have to implement
clone_ and borrow_ methods for a bunch of properties.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/properties/properties.mako.rs, components/style/animation.rs, components/style/properties/data.py

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style code, but no tests are modified. Please consider adding a test!

emilio committed:
style: Add a generic way to deal with lists of values, ditch all uses of as_servo in style/animations.rs

bholley commented:
I don't understand why we introduce need_borrow and then remove it again?

bholley commented:
(the reason I did clone instead of borrow is that borrowing will almost never be possible from geckolib, because we need to convert the types)

emilio commented:
Yup, I'm doing some changes, sorry for the early review request

emilio committed:
style: remove is_servo(), as_servo() and as_servo_mut()

This commit adds a need_index prop to the style system, and autogenerates
iterators, and a get_xxx_mod(i) function from a get_xxx_prop() and
get_xxx_at(index) functions.

This allows us to (finally!) take rid of the as_servo() hack. There are a few
unimplemented clones, but I'm just too lazy for that right now.

emilio commented:
Ok, now it should be ready for review.

r? @bholley

bholley commented on components/style/properties/longhand/box.mako.rs:
```diff
@@ -343,7 +344,9 @@ ${helpers.single_keyword("overflow-x", "visible hidden scroll auto",
-<%helpers:longhand name="transition-timing-function" animatable="False">
+<%helpers:longhand name="transition-timing-function"
+                   need_index="True"
+                   animatable="False">
```
Nit - the order here isn't consistent.

bholley commented on ports/geckolib/properties.mako.rs:
```diff
@@ -265,14 +293,20 @@ def set_gecko_property(ffi_name, expr):
 </%def>
 
-<%def name="impl_color(ident, gecko_ffi_name, color_flags_ffi_name=None)">
+<%def name="impl_color(ident, gecko_ffi_name, color_flags_ffi_name=None, need_clone=None)">
```
This should be False, not None.

emilio replied:
Ack'd

bholley commented on ports/geckolib/properties.mako.rs:
```diff
@@ -245,15 +246,42 @@ def set_gecko_property(ffi_name, expr):
     }
 </%def>
 
+<%def name="impl_color_setter(ident, gecko_ffi_name, color_flags_ffi_name=None)">
```
Looks like this is a duplicate of the above - remove it?

emilio replied:
Same reason, my fault

bholley commented on ports/geckolib/properties.mako.rs:
```diff
@@ -245,15 +246,42 @@ def set_gecko_property(ffi_name, expr):
 <%def name="impl_color_copy(ident, gecko_ffi_name, color_flags_ffi_name=None)">
     fn copy_${ident}_from(&mut self, other: &Self) {
         % if color_flags_ffi_name:
-        ${clear_color_flags(color_flags_ffi_name)}
```
Why are we removing the call to clear_color_flags?

emilio replied:
Whoops, good catch. At one point I wrote another version of these functions, but then backed it out, and this seems that it got lost in the meantime.

bholley commented on ports/geckolib/values.rs:
```diff
@@ -12,65 +12,147 @@ use style::values::computed::{LengthOrPercentage, LengthOrPercentageOrAuto, Leng
 }
 
-pub trait ToGeckoStyleCoord {
+pub trait ToGeckoStyleCoord : Sized {
```
Might want to rename this to GeckoStyleCoordConversion, or some potentially-more-idiomatic name that reflects the bidirectional conversion.

bholley commented:
r=me with those fixes.

@bors-servo delegate+

bors-servo commented:
:v: @emilio can now approve this pull request

emilio commented:
@bors-servo: r=bholley

bors-servo commented:
:pushpin: Commit 808f5ab has been approved by `bholley`

<!-- @bors-servo r=bholley 808f5ab75737c0da1d251735d672996ccdf397ad -->

bors-servo commented:
:hourglass: Testing commit 808f5ab with merge 540aeaa...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1838)

emilio commented:
These were legit failures (`get_mod` did `count % element` instead of the opposite). It's fixed.

@bors-servo: r=bholley

bors-servo commented:
:pushpin: Commit 30963dd has been approved by `bholley`

<!-- @bors-servo r=bholley 30963dd74da5e9a0e2fd7c8d56ec23aa474e8dd3 -->

bors-servo commented:
:hourglass: Testing commit 30963dd with merge 00407fa...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/1922)

highfive commented:
▶ FAIL [expected PASS] /css-transforms-1_dev/html/transform-table-007.htm
      └   → /css-transforms-1_dev/html/transform-table-007.htm a5c014b20ef1363bea6f24eda28c7efb7c45698a
    /css-transforms-1_dev/html/reference/transform-blank-ref.htm fa6407b1acbbfea27e27061e7d1bdeca98e4a728
    Testing a5c014b20ef1363bea6f24eda28c7efb7c45698a == fa6407b1acbbfea27e27061e7d1bdeca98e4a728

emilio commented:
@bors-servo: retry

 * https://github.com/servo/servo/issues/11574

bors-servo commented:
:hourglass: Testing commit 30963dd with merge aa8c23f...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1907)

emilio commented:
@bors-servo: retry

* iframe-parentage (I can't look up the issue number right now)
On Jul 1, 2016 5:33 PM, "bors-servo" <notifications@github.com> wrote:

> 💔 Test failed - mac-rel-wpt
> <http://build.servo.org/builders/mac-rel-wpt/builds/1907>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/servo/servo/pull/11972#issuecomment-230073283>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ABQwug_qu-cW3PRsOuvzDEi5iGo4Cqr7ks5qRbHdgaJpZM4JCoJm>
> .
>

bors-servo commented:
:hourglass: Testing commit 30963dd with merge f41828f...

bors-servo commented:
:boom: Test timed out

ConnorGBrewster commented:
@bors-servo retry
 * timeout?

bors-servo commented:
:hourglass: Testing commit 30963dd with merge bdfad79...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/1932)

highfive commented:
▶ FAIL [expected PASS] /css-transforms-1_dev/html/transform-table-007.htm
      └   → /css-transforms-1_dev/html/transform-table-007.htm a5c014b20ef1363bea6f24eda28c7efb7c45698a
    /css-transforms-1_dev/html/refe</span><span class="stdout">rence/transform-blank-ref.htm fa6407b1acbbfea27e27061e7d1bdeca98e4a728
    Testing a5c014b20ef1363bea6f24eda28c7efb7c45698a == fa6407b1acbbfea27e27061e7d1bdeca98e4a728

ConnorGBrewster commented:
@bors-servo retry
 * #11574

bors-servo commented:
:hourglass: Testing commit 30963dd with merge a77cc99...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/1934)

ConnorGBrewster commented:
@bors-servo retry
 * #11703

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/1683), [arm32](http://build.servo.org/builders/arm32/builds/1804), [arm64](http://build.servo.org/builders/arm64/builds/1794), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1828), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/1869), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1852), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1917), [windows](http://build.servo.org/builders/windows/builds/1530) are reusable. Rebuilding only [linux-rel](http://build.servo.org/builders/linux-rel/builds/1934)...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/1935)

jdm commented:
@bors-servo: retry
* #11703

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/1683), [arm32](http://build.servo.org/builders/arm32/builds/1804), [arm64](http://build.servo.org/builders/arm64/builds/1794), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1828), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/1869), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1852), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1917), [windows](http://build.servo.org/builders/windows/builds/1530) are reusable. Rebuilding only [linux-rel](http://build.servo.org/builders/linux-rel/builds/1935)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1683), [arm32](http://build.servo.org/builders/arm32/builds/1804), [arm64](http://build.servo.org/builders/arm64/builds/1794), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1828), [linux-rel](http://build.servo.org/builders/linux-rel/builds/1936), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/1869), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1852), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1917), [windows](http://build.servo.org/builders/windows/builds/1530)
